src/Parser/Desugar.ml

```diff
@@ -209,24 +209,26 @@ let rec tr_type_def (tp : Raw.type_expr) args =
   | TTypeLbl _ | TEffectLbl _ ->
     Error.fatal (Error.desugar_error tp.pos)
 
-let tr_ctor_decl (d : Raw.ctor_decl) =
+let tr_ctor_decl ~public (d : Raw.ctor_decl) =
   let make data = { d with data = data } in
   match d.data with
   | CtorDecl(name, { data = TRecord flds; _ } :: schs ) ->
     let (tvs, implicit) = map_inst_like tr_scheme_field flds in
     let schs = List.map tr_scheme_expr schs in
-    make (CtorDecl(name, tvs, implicit, schs))
+    make (CtorDecl(public, name, tvs, implicit, schs))
   | CtorDecl(name, schs) ->
-    make (CtorDecl(name, [], [], List.map tr_scheme_expr schs))
+    make (CtorDecl(public, name, [], [], List.map tr_scheme_expr schs))
 
 let tr_data_def (dd : Raw.data_def) =
   let make data = { dd with data = data } in
   match dd.data with
-  | DD_Data(tp, cs) ->
+  | DD_Data(vis, tp, cs) ->
+    let pub_type  = vis <> DV_Private in
+    let pub_ctors = vis =  DV_Public  in
```

Comment:
I'm not sure how this code is resistant to changes. What would happen, if we add new constructors to `data_vis` type? Will we find this place and change it accordingly, or just overlook it and introduce a bug to code?